src/auth/Crypto.cc

```diff
@@ -353,19 +365,405 @@ CryptoKeyHandler *CryptoAES::get_key_handler(const bufferptr& secret,
 {
   CryptoAESKeyHandler *ckh = new CryptoAESKeyHandler;
   ostringstream oss;
-  if (ckh->init(secret, oss) < 0) {
+  if (ckh->init(secret, bufferptr((unsigned)0), oss) < 0) {
     error = oss.str();
     delete ckh;
     return NULL;
   }
   return ckh;
 }
 
+CryptoKeyHandler *CryptoAES::get_key_handler(
+    const bufferptr& secret,
+    const bufferptr& iv, string& error)
+{
+  CryptoAESKeyHandler *ckh = new CryptoAESKeyHandler;
+    ostringstream oss;
+    if (ckh->init(secret, iv, oss) < 0) {
+      error = oss.str();
+      delete ckh;
+      return NULL;
+    }
+    return ckh;
+}
+
+class CryptoAES_256_CBC : public CryptoHandler {
+public:
+  CryptoAES_256_CBC() { }
+  ~CryptoAES_256_CBC() override {}
+  int get_type() const override {
+    return CEPH_CRYPTO_AES_256_CBC;
+  }
+  int create(bufferptr& secret) override;
+  int validate_secret(const bufferptr& secret) override;
+  CryptoKeyHandler *get_key_handler(const bufferptr& secret, string& error) override;
+  CryptoKeyHandler *get_key_handler(const bufferptr& secret, const bufferptr& iv, string& error) override;
+
+  static constexpr size_t KEY_SIZE = 256 / 8;
+  static constexpr size_t IV_SIZE = 128 / 8;
+  static constexpr size_t BLOCK_SIZE = 128 / 8;
+};
+
+
+class CryptoAES_256_CBCKeyHandler : public CryptoKeyHandler {
+public:
+  CryptoAES_256_CBCKeyHandler(const bufferptr& key, const bufferptr& iv)
+    : iv(iv) {
+    secret = key;
+  }
+  virtual ~CryptoAES_256_CBCKeyHandler() {}
+
+  int encrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return cbc_transform(in, out, true, error);
+  }
+  int decrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return cbc_transform(in, out, false, error);
+  }
+private:
+  bufferptr iv;
+  int cbc_transform(const bufferlist& in, bufferlist& out,
+      bool encrypt, std::string *error) const;
+  int cbc_transform(
+      const unsigned char* in, size_t size, unsigned char* out,
+      bool encrypt, std::string *error) const ;
+};
+
+int CryptoAES_256_CBCKeyHandler::cbc_transform(
+    const bufferlist& in, bufferlist& out,
+    bool encrypt, std::string *error) const
+{
+  if ((in.length() % CryptoAES_256_CBC::BLOCK_SIZE) != 0) {
+      return -EINVAL;
+    }
+  int result;
+  bufferptr p(in.length());
+  bufferlist _in(in);
+  result = cbc_transform(
+      reinterpret_cast<const unsigned char*>(_in.c_str()),
```

Comment:
i would really like to avoid calling `bufferlist::c_str()` here if we can avoid it, because it may have to reallocate to flatten the buffers (that's why `c_str()` isn't const)

would it work if you turned this into a for loop over `in.buffers()`, and called `cbc_transform()` with each `bufferptr`? the CryptoPP version of `CryptoAESKeyHandler` does something  similar with its encode/decode